File: it-agents/skills/code-architect/references/code-examples/typescript/State.ts
// State — let an object alter behavior when its internal state changes.

interface DoorState {
  open(door: Door): void;
  close(door: Door): void;
  label(): string;
}

class ClosedState implements DoorState {
  open(door: Door): void {
    console.log("opening door");
    door.setState(new OpenState());
  }
  close(): void {
    console.log("door already closed");
  }
  label(): string {
    return "closed";
  }
}

class OpenState implements DoorState {
  open(): void {
    console.log("door already open");
  }
  close(door: Door): void {
    console.log("closing door");
    door.setState(new ClosedState());
  }
  label(): string {
    return "open";
  }
}

// Context delegates behavior to the current state object.
class Door {
  private state: DoorState = new ClosedState();

  setState(state: DoorState): void {
    this.state = state;
  }
  open(): void {
    this.state.open(this);
  }
  close(): void {
    this.state.close(this);
  }
  get status(): string {
    return this.state.label();
  }
}

function demo(): void {
  const door = new Door();
  console.log(`status: ${door.status}`);
  door.open();
  door.open();
  door.close();
  console.log(`status: ${door.status}`);
}

demo();

File: it-agents/skills/code-architect/references/code-examples/go/state.go
// State: let an object change its behaviour when its internal state
// changes, by delegating to interchangeable state objects.
package main

import "fmt"

// State defines behaviour for one mode and may switch the context's state.
type State interface {
	Handle(c *context) string
}

// context delegates behaviour to its current state object.
type context struct {
	state State
}

func (c *context) Request() string { return c.state.Handle(c) }

type openState struct{}

func (openState) Handle(c *context) string {
	c.state = closedState{}
	return "was open -> now closed"
}

type closedState struct{}

func (closedState) Handle(c *context) string {
	c.state = openState{}
	return "was closed -> now open"
}

func main() {
	c := &context{state: openState{}}
	fmt.Println(c.Request())
	fmt.Println(c.Request())
	fmt.Println(c.Request())
}
